### PR TITLE
BUG FIX: Multipart/form-data crashed if data was missing

### PR DESCRIPTION
**Pull request by TheInkedEngineer**
Fix a crash in `FormDataDecoderContext` when the offset doesn't exist when parsing multipart form data (#2548)

### CODE REVIEW

File: Sources/Vapor/Multipart/FormDataDecoder.swift [hunk 1]
```diff
@@ -75,9 +75,12 @@ private final class FormDataDecoderContext {
         case 2:
             let name = codingPath[0].stringValue + "[]"
             guard let offset = codingPath[1].intValue else {
-                throw MultipartError.nesting
+              throw MultipartError.nesting
```
Comment (0xTim) [inline Sources/Vapor/Multipart/FormDataDecoder.swift:78]: Nit - style guide is 4 spaces
```suggestion
                throw MultipartError.nesting
```

File: Sources/Vapor/Multipart/FormDataDecoder.swift [hunk 2]
```diff
@@ -75,9 +75,12 @@ private final class FormDataDecoderContext {
...
             }
-            part = parts.allParts(named: name)[offset]
+            guard let p = parts.allParts(named: name)[safe: offset] else {
+              throw MultipartError.missingPart("\(codingPath[1].stringValue)")
```
Comment (0xTim) [inline Sources/Vapor/Multipart/FormDataDecoder.swift:81]: Nit - style guide is 4 spaces
```suggestion
                throw MultipartError.missingPart("\(codingPath[1].stringValue)")
```

**Review by 0xTim — CHANGES_REQUESTED** [review]
Couple of small style guide nits and we're good to go!

**Review by 0xTim — APPROVED** [review]
Thanks!

**Comment by TheInkedEngineer** [comment]
My bad, my Xcode default indentation is 2. I'll pay better attention next time.
Cheers

**Comment by 0xTim** [comment]
Original description:

This Pull request aims to do two things:

1- Fix an already existing bug: The `FormDataDecoderContext` was crash prone since it did not check if the offset exists.

2- Open a discussion about the problem that invoked this bug to begin with.

## The fix
Check if the part at the offset exists, otherwise throw an error `MultipartError.missingPart("\(codingPath[1].stringValue)")`

## The bigger problem

`URL` is decoded in the following manner:
```swift
extension URL : Codable {
    private enum CodingKeys : Int, CodingKey {
        case base
        case relative
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let relative = try container.decode(String.self, forKey: .relative)
        let base = try container.decodeIfPresent(URL.self, forKey: .base)

        guard let url = URL(string: relative, relativeTo: base) else {
            throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: decoder.codingPath,
                                                                    debugDescription: "Invalid URL string."))
        }

        self = url
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.relativeString, forKey: .relative)
        if let base = self.baseURL {
            try container.encode(base, forKey: .base)
        }
    }
}
```
Therefore generating 2 coding path.

However, the way is vapor is currently built, it assumes that when multiform data contain more than one `CodingPath` it's because it's an array, which as demonstrated above, is not always the case.

To replicate this case, this simple code should be sufficient:

```swift
      let data = """
      --hello\r\n\
      Content-Disposition: form-data; name="link"\r\n\
      \r\n\
      https://google.com\r\n\
      --hello--\r\n
      """

      struct Foo: Decodable {
        var link: URL
      }
```

And check the problem arise in the following part of `FormDataDecoderContext`:
```swift
        case 2:
            let name = codingPath[0].stringValue + "[]"
            guard let offset = codingPath[1].intValue else {
                throw MultipartError.nesting
            }
```

**Comment by tanner0101** [comment]
These changes are now available in [4.37.1](https://github.com/vapor/vapor/releases/tag/4.37.1)

**Pull request merged**